README.md
# Aquila MQTT-SN Gateway

MQTT-SN gateway for the Aquila 2.0 platform.

This software acts as a transparent link between a sensor network of low power devices (like [Altair](http://www.aquila.io/en) or other 802.15.4 or RF devices) and a MQTT broker (like mosca or mosquitto). This allows us to seamlessly and easily integrate those devices with existing MQTT applications and libraries.

You can find more information in the [documentation](doc/)

```
    (Device)_____                          ____________________            _______________
                  \___________            |                    |          |               |
                      MQTT-SN \__ ________|                    |   MQTT   |               |
          (Device)_______________| Bridge |      Gateway       |__________|  MQTT Broker  |
                               __ --------|  (aquila-gateway)  |          |               |
            __________________/           |                    |          |               |
    (Device)                               --------------------            ---------------
                                                                                  |
                                                                             MQTT |
                                                                           _______|________
                                                                          |               |
                                                                          |  Other MQTT   |
                                                                          |    Devices    |
                                                                           ---------------
```

## Bridge and client implementations

This gateway is meant to be as transport agnostic as possible, this allows us to use almost any sensor network just by developing a hardware bridge that implements the serial forwarder protocol defined in the [documentation](doc/).

Currently there are implementations for the [Altair](http://www.aquila.io/en) 802.15.4 development board and for rfm69 915Mhz RF devices with Atmel AVR processors.

- Altair: [Bridge](https://github.com/Rodmg/altair-mqtt-sn-bridge) and [example client template](https://github.com/Rodmg/altair-mqtt-sn-client-example)
- rfm69: [Bridge](https://github.com/Rodmg/rfm-mqtt-sn-bridge) and [example client template](https://github.com/Rodmg/rfm-mqtt-sn-client-example)

## Requirements

- [Node.js](https://nodejs.org/en/) v6.5.X or newer. (Tested with v6.10.2+)

## Usage

1. Install:

  ```
  npm install -g aquila-gateway bunyan
  ```

  *If you have problems installing, try with:* ``sudo npm install -g aquila-gateway bunyan --unsafe-perm`` 

2. Run a MQTT broker on your PC, for example [Mosca](https://github.com/mcollina/mosca)

3. Connect the Bridge to the PC and identify which serial port it's connected to

4. Run:

  ```
  aquila-gateway -p <your Bridge serial port> | bunyan
  ```

## Advanced usage

Get help:

```
aquila-gateway -h
```

```
Usage: aquila-gateway [options]

  Options:

    -h, --help                                output usage information
    -V, --version                             output the version number
    -v, --verbose [level]                     Verbosity level for logging (fatal, error, warn, info, debug, trace) [info]
    -t, --transport [transport]               Forwarder transport type (serial, tcp) [serial]
    -p, --port [serial port]                  Serial Port path if using serial transport, TCP port number if using TCP transport [/dev/tty.SLAB_USBtoUART | 6969]
    -b, --broker [url]                        MQTT broker URL [http://localhost:1883]
    -u, --allow-unknown-devices [true/false]  Allow connection of previously unknown (not paired) devices [true]
    -s, --subnet [pan id]                     PAN subnet number (1 to 254) [1]
    -k, --key [16 byte array]                 16 byte encryption key [null]
    -d, --data-path [path]                    Path to data persist file [/Users/rod/.aquila-gateway/data.json]
    -m, --monitor-prefix [prefix]             Gateway monitor topics prefix [gw]
```

Connect to a remote broker (example):

```
aquila-gateway -p /dev/tty.SLAB_USBtoUART -b http://test.mosquitto.org:1883 | bunyan
```

## Developement

1. Clone this repository and ``cd`` to the project directory

2. Install dependencies:

  ```
  npm install
  npm install -g bunyan gulp-cli
  ```
3. Run a MQTT broker on your PC, for example [Mosca](https://github.com/mcollina/mosca)

4. Connect the Bridge to the PC and identify which serial port it's connected to

5. Run:

  ```
  ./aquila-gateway.js -p <your Bridge serial port> | bunyan
  ```

6. Development tasks (uses gulp):

  **Build:**

  ```
  gulp build
  ```

  **Build and run (you may need to edit gulpfile.js for setting the serial port):**

  ```
  gulp serve | bunyan
  ```

  **Automatically watch for changes:**

  ```
  gulp watch | bunyan
  ```

## Supported MQTT-SN features

- QoS: supports QoS0, QoS1 and QoS2 (QoS2 implementation between device and gateway is mostly dummy, equivalent to QoS1)
- Publish retries for QoS1 and QoS2
- Commands:
  - ADVERTISE
  - CONNECT
  - CONNACK
  - DISCONNECT
  - WILLTOPICREQ
  - WILLTOPIC
  - WILLMSGREQ
  - WILLMSG
  - REGISTER
  - REGACK
  - PUBLISH
  - PUBACK
  - SUBSCRIBE
  - SUBACK
  - UNSUBSCRIBE
  - UNSUBACK
  - PINGREQ
  - PINGRESP
  - WILLTOPICUPD
  - WILLMSGUPD
  - WILLTOPICRESP
  - WILLMSGRESP
  - PUBREC, PUBREL, PBCOMP (QoS2)
  - SEARCHGW (basic, not spec checked)
  - GWINFO (basic, not spec checked)
- Implements Forwarder Encapsulation spec with a little modification: adds lqi and rssi data at the start of the frame.
- Supports topic register
- Supports last will
- Supports and manages disconnect timeout
- Will update
- Subscribe, Unsubscribe
- Retained messages support (issue: when a device subscribes to a topic with a retained message, it will be resent to all devices susbscribed to it, check if it's a problem)
- Sleeping nodes -> message buffering

## Extra features (not MQTT-SN standard)

- Gateway management via predefined topics for getting info of connected devices, management of the devices, etc.
- Device pairing management for Altair and rfm69 (Transport dependent)
- Multiple forwarder software transports support: Serial and TCP (experimental proof of concept) are currently implemented.

## TODO

- QoS -1
- WILDCARDS support as mqtt-sn spec
- Short and predefined MQTT-SN topics

## Not supported

- 3 byte MQTT-SN header

## TOTEST

- Check if parser on willtopicupd accept empty flags and topic (for removing will)
- Will update
- Make sure that buffered messages are sent in order


CHANGELOG.md

# Changelog

## develop ()

**Breaking Changes:**

- Changed database to sqlite with the sequelize library. This means that any data from previous versions **will be lost**.

## 0.6.0 (2017-02-13)

- Implemented removing devices from the monitor API.

- Changed Monitor API: All "*/get" command topics are now "*/req" for consistency with responses ("*/res"). Old API is still supported but deprecated.

- Minor bug fix: incorrect buffer function usage in Serial and TCP Transports.

The first change is useful when managing devices from a custom app. Usage:

Removing a device:

1. subscribe to gw/devices/remove/res
2. publish to gw/devices/remove/req a json with the address or id of the device to delete. Example: { "address": 23 } or { "id": 23 }.
3. You should get a JSON response via the subscription to gw/devices/remove/res: { "success": true } (or false).

More info in the docs: https://github.com/Rodmg/aquila-mqtt-sn-gateway/blob/master/doc/index.md

## 0.5.0 (2016-11-05)

- Fixed issues with sleeping devices getting lost.

- Added command line option for specifying path of the device database file.

- Added command line option for specifying the Gateway monitor topics prefix.

The last two changes are useful when running multiple gateway instances on the same machine, so that they don't interfere with each other.

## 0.4.2 (2016-10-22)

- Now you can install it directly from npm as a global package and use it as a command line program. Instructions are on the README of the github repo.

- Implemented experimental TCP Transport: Instead of connecting the Bridge directly via Serial, you can connect it remotely via a TCP transport. For example, connecting the Bridge to a ESP8266 and then to the Gateway via WiFi.

- Framing error fixes: The serial communication between the Bridge and the Gateway uses a protocol called SLIP, it has an start and end character for delimiting packets. Sometimes, when starting the Gateway, the Bridge could be in the middle of receiving a packet from a device and the Gateway would get an incomplete packet, this would cause a "Framing error". Ideally, it should only lose that packet and continue normally, but due the way the node-slip library was implemented, it would cause all the following packets to be also erroneous, a "frame error train".

I added the validation for that special case to the node-slip library in a fork and used it in aquila-gateway, now it works as expected.

- Other minor bug fixes.

**Embedded MQTT-SN library - MQTTClient:**

- Fixed a bug: When connecting with "last will" enabled, it would always return a success even when it failed.

- Updated examples, changed default PAN address to 1, the default in aquila-gateway.

src/main.ts

import { TCPTransport } from './TCPTransport';
import { SerialTransport } from './SerialTransport';
import { MQTTTransport } from './MQTTTransport';
import { GatewayDB } from './GatewayDB';
import { Forwarder } from './Forwarder';
import { Gateway } from './Gateway';
import { GwMonitor } from './GwMonitor';
import { log } from './Logger';
import * as program from 'commander';
import * as path  from 'path';
const pjson = require('./../package.json');

function parseBool(s: string)
{
  if(s === 'false') return false;
  return true;
}

function parseKey(key: string)
{
  let keyArr: Array<any> = key.split(',');
  if(keyArr.length !== 16)
  {
    log.warn("Invalid encryption key received, starting without encryption");
    return null;
  }
  keyArr = keyArr.map((item) => {
    return parseInt(item);
  });
  return keyArr;
}

function parseSubnet(s: string)
{
  return parseInt(s);
}

const DEFAULT_DATA_DIR = path.join(process.env[(process.platform === 'win32') ? 'ALLUSERSPROFILE' : 'HOME'], '.aquila-gateway');
const DEFAULT_DATA_PATH = path.join(DEFAULT_DATA_DIR, 'data.json');

program
  .version(pjson.version)
  .option('-v, --verbose [level]', 'Verbosity level for logging (fatal, error, warn, info, debug, trace) [info]', 'info')
  .option('-t, --transport [transport]', 'Forwarder transport type (serial, tcp) [serial]', 'serial')
  .option('-p, --port [serial port]', 'Serial Port path if using serial transport, TCP port number if using TCP transport [/dev/tty.SLAB_USBtoUART | 6969]', '/dev/tty.SLAB_USBtoUART')
  .option('-b, --broker [url]', 'MQTT broker URL [http://localhost:1883]', 'http://localhost:1883')
  .option('-u, --allow-unknown-devices [true/false]', 'Allow connection of previously unknown (not paired) devices [true]', parseBool, true)
  .option('-s, --subnet [pan id]', 'PAN subnet number (1 to 254) [1]', parseSubnet, 1)
  .option('-k, --key [16 byte array]', '16 byte encryption key [null]', parseKey, null)
  .option('-d, --data-path [path]', 'Path to data persist file [' + DEFAULT_DATA_PATH + ']', DEFAULT_DATA_PATH)
  .option('-m, --monitor-prefix [prefix]', 'Gateway monitor topics prefix [gw]', 'gw')
  .parse(process.argv);

log.level(program.verbose);

// Select Forwarder transport
let transport;
if(program.transport === 'tcp')
{
  let tcpPort = parseInt(program.port);
  if(isNaN(tcpPort)) tcpPort = 6969;
  transport = new TCPTransport(tcpPort);
}
else if(program.transport === 'mqtt')
{
  transport = new MQTTTransport(program.broker, 
    "91bbef0fa64c130d0b274c7299c424/bridge/in", 
    "91bbef0fa64c130d0b274c7299c424/bridge/out");
}
else
{
  transport = new SerialTransport(115200, program.port);
}

let db = new GatewayDB(program.dataPath);
let gw: Gateway;

db.connect()
.then(() => {
  let forwarder = new Forwarder(db, transport, program.subnet, program.key);
  gw = new Gateway(db, forwarder);
  return gw.init(program.broker, program.allowUnknownDevices);
})
.then(() => {
  let gwMon = new GwMonitor(gw, program.monitorPrefix);
  log.info("Gateway Started");
});


